### ISSUE TITLE
Refinement of "ignore unknown fields" guideline

### ISSUE THREAD

**Issue by tfrauenstein** [zalando/restful-api-guidelines#165]
We currently have the general rule "Should: Prefer Compatible Extensions" and the more client specific rule "Must: Prepare Clients for Compatible API Extensions (the Robustness Principle)".
Both are partially overlapping -- I think we should specialise the first to be server specific and avoid redundancies and even inconsistencies. 

For the server side, I'm not sure whether here robustness principle generally applies; there are some problems, for instance...
a) might not be compatible with semantics of operation, e.g. PUT for resource storage service should not ignore unknown input fields and do not store it without any feedback 
b) ignore unknown input fields may have deferred impact to future extensions which are supposed to be compatible, however, might break clients that use a different type here

Having this in mind, I propose on server side "ignore unknown fields" is not a "should", but a "could" rule.
If robustness will not be provided, I would expect that still additional properties is true -- because I want to have the option to extend the data structure in the future (see issue #164 ) -- but a generic error code (which one?) is returned. 

Yes, different services might want to take different approaches with "unknown fields" -- but the service provider should be aware of the problems mentioned above and be explicit if it is not supported.

**Comment by tfrauenstein** [comment]
short summary what we discussed today in Guild meeting:
- client and server view must be clearly segregated
- clients must ignore unknown fields in servers output data 
- when server ignores unknown fields in its input data then it has the following effects:
  - PUT and GET may not be symmetric, unknown fields can be ignored by PUT will not be visible subsequent in GET output
  - certain client errors cannot be recognized by server, e.g. typing error for optional fields will be interpreted as additional unknown field ignored instead of corresponding server error feedback
  - future extensions of the input data structure will not be compatible, i.e. break clients that already use this field but with different type
- servers might want to take different approaches with "unknown fields" -- but the service provider should be aware of the problems mentioned above and be explicit in what is supported
- discussed yet w/o consensus: should we recommend servers to ignore or not to ignore unknown fields

**Comment by whiskeysierra** [comment]
> certain client errors cannot be recognized by server, e.g. typing error for optional field names will be interpreted as additional unknown field ignored instead of corresponding server error feedback

Why?

**Comment by fmueller** [comment]
For this case @whiskeysierra, do you mean server should recognise typing errors in field names?

**Comment by whiskeysierra** [comment]
I understand *typing errors* as using the correct field name but the wrong data type. Are you talking about typos in the field names?

**Comment by tfrauenstein** [comment]
yes, I mean "field name typing errors" -- server cannot differ between client intensional provided additional field vs. client field name typing error where he intensionally wanted to provide an optional input field.

**Comment by whiskeysierra** [comment]
OK, got it.

**Comment by dehora** [comment]
> short summary what we discussed today in Guild meeting:

> - client and server view must be clearly segregated

👍 

> - clients must ignore unknown fields in servers output data 

👍 . We'll want to clarify wording slightly - "must ignore" as intended means "must not crash", but it's ok for generic processing engines to handle generic json (ie it's not _required_ that clients drop data)

> - when server ignores unknown fields in its input data then it has the following effects:

>  - PUT and GET may not be symmetric, unknown fields can be ignored by PUT and will not be visible subsequent in GET output

This makes sense for POST/PATCH, but PUT means replace and HTTP's pretty clear about the default roundtrip expectations [RFC7231 section 4.3.4](https://tools.ietf.org/html/rfc7231#section-4.3.4): 

> A successful PUT of a given
   representation would suggest that a subsequent GET on that same
   target resource will result in an equivalent representation being
   sent in a 200 (OK) response.  However, there is no guarantee that
   such a state change will be observable, since the target resource
   might be acted upon by other user agents in parallel, or might be
   subject to dynamic processing by the origin server, before any
   subsequent GET is received. 

But as is typical with HTTP, it leaves enough room for some vagueness ("dynamic processing" 😞 ). Anyway, I think we'll want clear guidance here if PUT is allowed to be overloaded to means partial/selective storage/update.

>  - certain client errors cannot be recognized by server, e.g. typing error for optional field names will be interpreted as additional unknown field ignored instead of corresponding server error feedback

👍 This isn't actionable client/server guidance at runtime, but useful because it clarifies who's responsible for issues.

>  - future extensions of the input data structure will not be compatible, i.e. will break clients that already use this field but with different type

👍 

> - servers might want to take different approaches with "unknown fields" -- but the service provider should be aware of the problems mentioned above and be explicit in what is supported

👍 This is relevant to event processing or any scenario where there's long term storage and you're expected to be able to read historical data back out. I think where this is critical we can consider supporting team with media type formats that have clearer semantics than JSON Schema and canonical json mappings (eg Avro/protobuf3)

> - discussed yet w/o consensus: should we recommend servers to ignore or not to ignore unknown fields

I think it's best to be clear on the tradeoffs in the guidance as per the suggestions above, and provide guidance on how to clearly signal write-side validation errors maybe with standard Problem URIs, and document must-ignore style processing for POST/PATCH/PUT, that would result in data being dropped by the server.

**Comment by codekultur** [comment]
I totally agree, that clients should ignore unknown fields.

When it comes to the server side, I think we should introduce a rule, that server should (or event must) NOT ignore unkown fields (in PUT, POST, PATCH). IMHO the server defines the API and API enhancements are driven by the server side. So I think the server should introduce new fields first, before clients are allowed to use them. API enhancements are not driven by the clients by just introducing new fields. It is driven by developers of clients talking to the team providing the API.

Typos or scenarios where a client app thinks that certain fields are stored, but the server silently ignores these fields, are dangerous and very hard to detect. So I would argue, that unknown fields should not be ignored.

**Comment by codekultur** [comment]
There might be one use case where the server should accept unknown fields: When the server side is a generic storage engine. But in this case the server usually makes no assumptions about the format (or semantics) of the stored data. So I would not expect that the only data type defined in the API spec is a string, that then might contain any data structure, that is opaque to the server.

**Comment by tfrauenstein** [comment]
fixed by PR #248
